services/containerregistry/refreshtoken.go

```diff
@@ -0,0 +1,112 @@
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// ExchangeRefreshToken exchanges an Azure AD token for an Azure Container Registry refresh token
+func (c *ContainerRegistryClient) ExchangeRefreshToken(ctx context.Context) (string, *RefreshTokenClaims, error) {
+	token, err := c.authorizer.Token()
+	if err != nil {
+		return "", nil, err
+	}
+
+	serviceURL, err := parseService(c.serverURL)
+	if err != nil {
+		return "", nil, err
+	}
+
+	data := url.Values{}
+	data.Set("grant_type", "access_token")
+	data.Set("service", serviceURL.Hostname())
+	data.Set("access_token", token.AccessToken)
+	if len(c.tenantID) > 0 {
+		data.Set("tenant", c.tenantID)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, fmt.Sprintf("%s://%s/oauth2/exchange", serviceURL.Scheme, serviceURL.Host), strings.NewReader(data.Encode()))
+	if err != nil {
+		return "", nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	res, err := c.httpClient.Do(req)
+	if err != nil {
+		return "", nil, err
+	}
+
+	resBytes, err := io.ReadAll(res.Body)
+	if err != nil {
+		return "", nil, err
+	}
+
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return "", nil, fmt.Errorf("received non-200 status code - %d: %s", res.StatusCode, string(resBytes))
+	}
+
+	var resData struct {
+		RefreshToken string `json:"refresh_token"`
+	}
+
+	err = json.Unmarshal(resBytes, &resData)
+	if err != nil {
+		return "", nil, err
+	}
+
+	rtClaims, err := decodeRefreshTokenWithoutValidation(resData.RefreshToken)
+	if err != nil {
+		return "", nil, err
+	}
+
+	return resData.RefreshToken, rtClaims, nil
+}
+
+// RefreshTokenClaimsPermissions contains the actions permitted for the refresh token
+type RefreshTokenClaimsPermissions struct {
+	Actions    []string `json:"Actions"`
+	NotActions []string `json:"NotActions"`
+}
+
+// RefreshTokenClaims contians the claims of the refresh token
+// Please observe: These claims are in no way validated, only decoded
+type RefreshTokenClaims struct {
+	JwtID          string                        `json:"jti"`
+	Subject        string                        `json:"sub"`
+	NotBefore      int64                         `json:"nbf"`
+	ExpirationTime int64                         `json:"exp"`
+	IssuedAt       int64                         `json:"iat"`
+	Issuer         string                        `json:"iss"`
+	Audience       string                        `json:"aud"`
+	Version        string                        `json:"version"`
+	Rid            string                        `json:"rid"`
+	GrantType      string                        `json:"grant_type"`
+	ApplicationID  string                        `json:"appid"`
+	Permissions    RefreshTokenClaimsPermissions `json:"permissions"`
+	Roles          []string                      `json:"roles"`
+}
+
+func decodeRefreshTokenWithoutValidation(token string) (*RefreshTokenClaims, error) {
```

Comment:
As above, the auth package already has a `ParseClaims` function.

Reply:
I agree! Will look at it 👍

Reply:
As by the previous comment, if we're not using the same type for the Claims, we won't be able to use the same parse function until go 1.18 and generics.

Reply:
Sounds good, as above I think parsing refresh tokens should be a common feature of the `auth` package to avoid reimplementing it for subsequent services.